### PR TITLE
chore: replace atty with is-terminal

### PR DESCRIPTION
**Pull request by jcgruenhage**
`atty` is unmaintained and has a potential unaligned read. See https://github.com/rustsec/advisory-db/blob/main/crates/atty/RUSTSEC-2021-0145.md.

`is-terminal` is a replacement based on `atty`, with the soundness issue fixed and an (IMO) nicer to use API, mirroring what's available in the std lib on nightly with `std::io::IsTerminal`.

### CODE REVIEW

**Review by lemmih — APPROVED** [review]

**Comment by lemmih** [comment]
@jcgruenhage The `is-terminate` crate somehow requires nightly when targeting wasi32. Do you know why?

**Comment by jcgruenhage** [comment]
No clue, sorry. What's the error message you're getting? As far as I can tell, `is-terminal` should work one any Rust version >= 1.48.0 and not require nightly.

**Comment by lemmih** [comment]
> No clue, sorry. What's the error message you're getting? As far as I can tell, `is-terminal` should work one any Rust version >= 1.48.0 and not require nightly.

It might use `libc` with a different set of features which then cause trouble in the `errno` crate. Will investigate.

**Comment by nbarrios1337** [comment]
> @jcgruenhage The `is-terminate` crate somehow requires nightly when targeting wasi32. Do you know why?

Did some digging on the wasi32 compilation error:

- `is-terminal` v0.4.3 depends on `rustix` v0.36.8, leading to a transitive dependency on `errno` v0.2.8, resulting in [this PR's CI error](https://github.com/bheisler/criterion.rs/actions/runs/3627790596/jobs/6118598614#step:12:40).
- `errno` recently [merged a PR that removed use of `feature(thread_local)`](https://github.com/lambda-fairy/rust-errno/pull/66) and [released v0.3](https://crates.io/crates/errno/0.3.0) that encapsulates that change among others
- `rustix` recently [merged a PR that bumps its errno dependency to v0.3.0](https://github.com/bytecodealliance/rustix/pull/536)

To summarize what this PR needs:

- `is-terminal` would have to update its `rustix` dependency once `rustix` releases a version with the `errno` version bump change
- This PR would then need to update to that `is-terminal` version in order for it to not fail on the aforementioned CI error

**Comment by bgianfo** [comment]
> To summarize what this PR needs:
> 
> * `is-terminal` would have to update its `rustix` dependency once `rustix` releases a version with the `errno` version bump change
> * This PR would then need to update to that `is-terminal` version in order for it to not fail on the aforementioned CI error

Looks like `is-terminal` merged an update to it's `rustix` version, but hasn't quite cut a new release yet. https://github.com/sunfishcode/is-terminal/pull/20

**Comment by nickelc** [comment]
A new version has just been released.
https://crates.io/crates/is-terminal/0.4.6

**Comment by jcgruenhage** [comment]
@nickelc thanks for the notification, I've pulled in that release and rebased on latest master. Hopefully that should be merge-able now?

**Comment by nbarrios1337** [comment]
Looks like the `csv` crate updated their MSRV to rustc 1.60 as of csv  v1.2.0. From my understanding, criterion would either need to (a) Bump its own MSRV to 1.60 or (b) set a stricter version requirement for its `csv` dependency (i.e. `~1.1` in the Cargo.toml)

**Comment by lemmih** [comment]
> Looks like the `csv` crate updated their MSRV to rustc 1.60 as of csv v1.2.0. From my understanding, criterion would either need to (a) Bump its own MSRV to 1.60 or (b) set a stricter version requirement for its `csv` dependency (i.e. `~1.1` in the Cargo.toml)

Bumping the MSRV would be the easiest solution. Anyone up for that task?

**Comment by nbarrios1337** [comment]
PR made to bump the MSRV: #665

**Comment by lemmih** [comment]
Hmm, it appears that `rustix` does not support wasm when using stable rust.

**Comment by sunfishcode** [comment]
Ah, that's a bug in rustix. I've submitted https://github.com/bytecodealliance/rustix/pull/589 to fix it.

**Comment by sunfishcode** [comment]
The wasm32-wasi build on stable Rust is now fixed in rustix 0.37.7 and 0.36.12.

**Comment by lemmih** [comment]
> The wasm32-wasi build on stable Rust is now fixed in rustix 0.37.7 and 0.36.12.

Nice!

**Comment by nbarrios1337** [comment]
Perhaps a small commit for the CI to rerun is unpinning the `tempfile` dep? I believe it was pinned in #665 because of the same rustix bug.

**Comment by jcgruenhage** [comment]
I just re-run the CI by rebasing and force-pushing instead. CI seems happy now!

**Comment by dianpopa** [comment]
It s great to see this merged! Is it possible to also create an official tag for the change? (I expect 0.4.1)? 
We use this in [Firecracker](https://github.com/firecracker-microvm/firecracker/issues/3284) for benchmarking our versionize functionality. 
Thanks!

**Comment by lemmih** [comment]
New releases can only be made by @bheisler.

**Comment by dianpopa** [comment]
@bheisler can you please release a new version that includes this fix? Thanks!

**Comment by lemmih** [comment]
> @bheisler can you please release a new version that includes this fix? Thanks!

I have permission to release a new version. Will need to review this PR first (https://github.com/bheisler/criterion.rs/pull/679) and then update the CHANGELOG. A new is coming but it'll come faster with help.
